src/wallet/wallet2.cpp

```diff
@@ -3136,13 +3136,24 @@ bool wallet2::store_keys(const std::string& keys_file_name, const epee::wipeable
   crypto::chacha20(account_data.data(), account_data.size(), key, keys_file_data.iv, &cipher[0]);
   keys_file_data.account_data = cipher;
 
-  unlock_keys_file();
+  std::string tmp_file_name = keys_file_name + ".new";
   std::string buf;
   r = ::serialization::dump_binary(keys_file_data, buf);
-  r = r && epee::file_io_utils::save_string_to_file(keys_file_name, buf); //and never touch wallet_keys_file again, only read
-  CHECK_AND_ASSERT_MES(r, false, "failed to generate wallet keys file " << keys_file_name);
+  r = r && epee::file_io_utils::save_string_to_file(tmp_file_name, buf);
+  if (!r) {
+      MERROR("Failed to generate wallet keys file " << tmp_file_name);
+      return false;
+  }
+
+  unlock_keys_file();
+  std::error_code e = tools::replace_file(tmp_file_name, keys_file_name);
```

Comment:
Doesn't seem to remove the original file, so the rename will fail (typically).

Reply:
Interesting. I did not know it worked like this. However, we're using std::rename, which says it's undefined (presumably because Windows then I guess).

Reply:
Hmm, I didn't realize that. I don't have Windows to test on, unfortunately, but like I said, the cache file is [already handled this way](https://github.com/monero-project/monero/blob/1e74586ee99e4bd89626d2eb4d23883cd91f0f81/src/wallet/wallet2.cpp#L4872) without removing the destination file, and that must be working on Windows, right?

FWIW, it actually seems to say [implementation-defined](http://www.cplusplus.com/reference/cstdio/rename/), not undefined, so it should be ok to rely on if it does in fact work.

If you do insist on changing it, the correct place to do that would be in `tools::replace_file()` itself, right?

Reply:
(The only other place it seems to be used, [here](https://github.com/monero-project/monero/blob/1e74586ee99e4bd89626d2eb4d23883cd91f0f81/src/cryptonote_core/cryptonote_core.cpp#L1634), doesn't remove the destination file first, either)

Reply:
Windows has a custom code path, so we can ignore it. See my last comment below.